packages/cli/src/utils/typeUtils.ts

```diff
@@ -9,3 +9,9 @@ export type StringForUnion = string & Record<never, never>;
 export type StaticArray = `${StaticAbiType}[${number}]`;
 // static arrays and inferred enum names get mixed together - this helper separates them
 export type ExtractUserTypes<UnknownTypes extends StringForUnion> = Exclude<UnknownTypes, AbiType | StaticArray>;
+
+// When type inference sees multiple uses of 1 generic, it can only guess
+// which of those are supposed to define the generic (and it will be wrong in complex situations).
+// This helper explicitly makes a type that's dependent on some generic,
+// and will not be inferred as the generic's definition.
+export type AsDependent<T> = T extends infer P ? P : never;
```

Comment:
What's an example for this?

Reply:
check out [parseStoreConfig.test-d.ts](https://github.com/latticexyz/mud/pull/509/files#diff-c9b49da8abbb495550842f78d3966ae8959e6316d04eeaaf62cbc3ae026bab54)